### DISCUSSION TITLE
[.pio/build/firmware/src/tasmota.ino.cpp.o] Error 1

### DISCUSSION THREAD

**Discussion by tobibachert** [benzino77/tasmocompiler#178] [Q&A] [answered]
Hi,

I hope this is the right place to post problems, please forgive me if it is not and maybe tell me how I could get support in another way.
I'm trying to compile a tasmota build for ESP32 with Tasmocompiler running in gitpod following [this link](https://gitpod.io/#https://github.com/benzino77/tasmocompiler) from the readme.md. 
I'm doing the following settings in TasmoCompiler 5.1.1:
- Wifi settings: empty
- Features: Hardware Generic ESP32, checked features: bluetooth, light sensors, air/gas sensors, scripts, temperature/humidity sensors, timers, web interface (probably I'm not getting all the names correct, I'm running it with german language)
- additional parameters: empty
- Version: tasmota version: v9.4.0 or development, tried both but makes no difference for the result, language german

This is the result (without dependencies to save lines, I saw in the preview that otherwise the post will be cut) - I have no idea what that means or what I could do to prevent this error.
Any help is highly appreciated!

Best regards, Tobias

Processing firmware (framework: arduino; platform: espressif32 @ 3.2.0; board: esp32dev)
--------------------------------------------------------------------------------
Platform Manager: Installing espressif32 @ 3.2.0
Platform Manager: espressif32 @ 3.2.0 has been installed!
The platform 'espressif32 @ 3.2.0' has been successfully installed!
The rest of the packages will be installed later depending on your build environment.
Tool Manager: Installing platformio/toolchain-xtensa32 @ ~2.50200.0
Tool Manager: toolchain-xtensa32 @ 2.50200.97 has been installed!
Tool Manager: Installing platformio/framework-arduinoespressif32 @ ~3.10006.0
Tool Manager: framework-arduinoespressif32 @ 3.10006.210326 has been installed!
Tool Manager: Installing platformio/tool-esptoolpy @ ~1.30000.0
Tool Manager: tool-esptoolpy @ 1.30000.201119 has been installed!
Tool Manager: Installing platformio/tool-mklittlefs @ ~1.203.200522
Tool Manager: tool-mklittlefs @ 1.203.210203 has been installed!
Tool Manager: Installing platformio/tool-scons @ ~4.40100.2
Tool Manager: tool-scons @ 4.40100.2 has been installed!
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (3.2.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 80MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30000.201119 (3.0.0) 
 - tool-mklittlefs 1.203.210203 (2.3) 
 - toolchain-xtensa32 2.50200.97 (5.2.0)
Converting tasmota.ino
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ strict
Found 102 compatible libraries
Scanning dependencies...
Dependency Graph

--- skipped ----

Building in release mode
*** use provided user_config_override.h as planned ***
*** use provided platformio_override.ini as planned ***
Use manually specified: COM5
Compiling .pio/build/firmware/src/StackThunk_light.cpp.o
Compiling .pio/build/firmware/src/WiFiClientSecureLightBearSSL.cpp.o
Compiling .pio/build/firmware/src/homekit.c.o
Compiling .pio/build/firmware/src/tasmota.ino.cpp.o
Generating partitions .pio/build/firmware/partitions.bin
Compiling .pio/build/firmware/libbc3/WiFi/ETH.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFi.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiAP.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiClient.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiGeneric.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiMulti.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiSTA.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiScan.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiServer.cpp.o
Compiling .pio/build/firmware/libbc3/WiFi/WiFiUdp.cpp.o
Compiling .pio/build/firmware/lib779/WiFiClientSecure/WiFiClientSecure.cpp.o
Compiling .pio/build/firmware/lib779/WiFiClientSecure/ssl_client.cpp.o
Compiling .pio/build/firmware/lib79e/HTTPClient/HTTPClient.cpp.o
Compiling .pio/build/firmware/libc1e/FS/vfs_api.cpp.o
Compiling .pio/build/firmware/libc1e/FS/FS.cpp.o
Compiling .pio/build/firmware/libff4/WebServer/Parsing.cpp.o
Compiling .pio/build/firmware/libff4/WebServer/WebServer.cpp.o
Compiling .pio/build/firmware/libff4/WebServer/detail/mimetable.cpp.o
Compiling .pio/build/firmware/lib4a2/Update/HttpsOTAUpdate.cpp.o
Compiling .pio/build/firmware/lib4a2/Update/Updater.cpp.o
/tmp/Tasmota/tasmota/xdrv_10_scripter.ino:84:17: note: #pragma message: universal file system used
 #pragma message "universal file system used"
                 ^
Compiling .pio/build/firmware/lib329/HTTPUpdate/HTTPUpdate.cpp.o
Compiling .pio/build/firmware/lib04b/ESPmDNS/ESPmDNS.cpp.o
Compiling .pio/build/firmware/lib60e/ESP32-to-ESP8266-compat/ESP32Wifi.cpp.o
Archiving .pio/build/firmware/lib779/libWiFiClientSecure.a
Archiving .pio/build/firmware/libbc3/libWiFi.a
Indexing .pio/build/firmware/lib779/libWiFiClientSecure.a
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino: In function 'void CCS811Detect()':
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino:84:6: error: redefinition of 'void CCS811Detect()'
 void CCS811Detect(void)
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811.ino:47:6: note: 'void CCS811Detect()' previously defined here
 void CCS811Detect(void)
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino: In function 'void CCS811Update()':
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino:103:6: error: redefinition of 'void CCS811Update()'
 void CCS811Update(void)  // Perform every n second
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811.ino:57:6: note: 'void CCS811Update()' previously defined here
 void CCS811Update(void)  // Perform every n second
      ^
Compiling .pio/build/firmware/lib60e/ESP32-to-ESP8266-compat/esp8266toEsp32.cpp.o
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino: At global scope:
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino:247:28: error: redefinition of 'const char HTTP_SNS_CCS811 []'
 const char HTTP_SNS_CCS811[] PROGMEM =
                            ^
/tmp/Tasmota/tasmota/xsns_31_ccs811.ino:84:12: note: 'const char HTTP_SNS_CCS811 [55]' previously defined here
 const char HTTP_SNS_CCS811[] PROGMEM =
            ^
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino: In function 'void CCS811Show(bool)':
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino:253:6: error: redefinition of 'void CCS811Show(bool)'
 void CCS811Show(bool json)
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811.ino:88:6: note: 'void CCS811Show(bool)' previously defined here
 void CCS811Show(bool json)
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino: In function 'bool Xsns31(uint8_t)':
/tmp/Tasmota/tasmota/xsns_31_ccs811_v2.ino:309:6: error: redefinition of 'bool Xsns31(uint8_t)'
 bool Xsns31(uint8_t function)
      ^
/tmp/Tasmota/tasmota/xsns_31_ccs811.ino:108:6: note: 'bool Xsns31(uint8_t)' previously defined here
 bool Xsns31(uint8_t function)
      ^
Indexing .pio/build/firmware/libbc3/libWiFi.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/aead/ccm.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/aead/eax.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/aead/gcm.c.o
Archiving .pio/build/firmware/libc1e/libFS.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/ccopy.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec16be.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec16le.c.o
Indexing .pio/build/firmware/libc1e/libFS.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec32be.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec32le.c.o
Archiving .pio/build/firmware/lib79e/libHTTPClient.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec64be.c.o
Indexing .pio/build/firmware/lib79e/libHTTPClient.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/dec64le.c.o
Archiving .pio/build/firmware/lib4a2/libUpdate.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc16be.c.o
Indexing .pio/build/firmware/lib4a2/libUpdate.a
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc16le.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc32be.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc32le.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc64be.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/enc64le.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/pemdec.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/codec/pemenc.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/ec/ec_all_m15.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/ec/ec_c25519_i15.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/ec/ec_curve25519.c.o
Compiling .pio/build/firmware/lib50a/bearssl-esp8266/ec/ec_default.c.o
*** [.pio/build/firmware/src/tasmota.ino.cpp.o] Error 1
========================= [FAILED] Took 36.38 seconds =========================

Environment    Status    Duration
-------------  --------  ------------
firmware       FAILED    00:00:36.382
==================== 1 failed, 0 succeeded in 00:00:36.382 ====================
Finished. Exit code: 1.

**Comment by drhideg** [comment, marked as answer]
It's looks like ESP32 use `USE_CCS811_V2` flag by default, and when you selecet air/gas sensors its add `USE_CCS811` which causes a collision while compile. At the moment you can add `#undef USE_CCS811` to custom parameters for a succesfull compilation.
I don't have ESP32 based device and CCS811 for test it.

**Reply by tobibachert** [reply]
Wow...Yes, you're right, with this addition in the custom parameters I can compile the firmware!
I will need to test later on my ESP32 development board if it's working, I'm not at home at the moment.

Thank you very much!

**Reply by benzino77** [reply]
Last changes fix this problem. You don't have to do `#undef ` until you want to use V2 version of CCS811.
There was a problem with base environment on which every ESP32 board is defined.

**Reply by benzino77** [reply]
To use V2 version of CCS811 put this in the Custom parameters text field:
```C
#undef USE_CCS811
#ifdef USE_CCS811_V2
  #undef USE_CCS811_V2
#endif
#define USE_CCS811_V2
```

**Comment by christophhhh** [comment]
I'm having the same problem. Where is that "Custom parameters text field" where I should paste the code above? I tried `user_config_override.h`, but that didn't work?
